### PR TITLE
Coupons: Fix issues listing search results on category selector

### LINKED ISSUE #6839: [Coupons] - Selecting a word from keyboard suggestion display results not found on Edit Product Categories screen
**Describe the bug**
On the Edit Product Categories screen, when searching and selecting a word from the keyboard suggestion, it will result in results not found. It could be because of the extra space on the selected word.

**To Reproduce**
1. Turn on Coupon Management experimental feature
2. Select a coupon
3. Edit the coupon code and go to Edit Product Categories
4. Search for a product category but use the keyboard suggestion to select the word
5. See that no results are found
6. Type the same word and see that results are displayed

**Screenshots**

https://user-images.githubusercontent.com/17252150/168041734-231b84fb-35b3-40ee-b31f-4e475a60f617.MP4

**Expected behavior**
Search should return the correct results when user selects a word from the suggestion

**Mobile Environment**
Device: iPhone 13
iOS version: 15.3.1
WooCommerce iOS version: AppCenter build pr6809-84ae456

### LINKED ISSUE #6840: [Coupons] - Can't search for subcategories
**Describe the bug**
On Edit Product Categories, looks like we can't search for subcategories, it will return results not found.

**To Reproduce**
1. Turn on Coupon Management experimental feature
2. Select a coupon
3. Edit the coupon code and go to Edit Product Categories
4. On search, search for a subcategory
5. See that it will show results not found

**Screenshots**

https://user-images.githubusercontent.com/17252150/168043287-b017cc32-346f-49c1-825e-5f122009317c.MP4

**Expected behavior**
We should be able to search for both categories and subcategories using the search bar

**Mobile Environment**
Device: iPhone 13
iOS version: 15.3.1
WooCommerce iOS version: AppCenter build pr6809-84ae456

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #6840 
Also closes: #6839 

### Description

Due to the small size of changes, this PR fixes a few issues of the search functionality on the product category selector screen:
- Fix the issue displaying subcategories on search results. We were ignoring the subcategories without parents on the list, I added another helper method to display all results disregarding the parent-child relationships. This is used only when displaying search results.
- Fix the incorrect predicate for searching categories. Previously there were missing brackets which made the filter for `siteID` being ignored. This was fixed with extra brackets around the check for matching search queries.
- Fix the issue searching with queries that have space around them. This was fixed by trimming space on both ends of the queries.

### Testing instructions

- Navigate to the edit screen of any coupon and select the edit categories button.
- Enter a keyword in the search bar that matches at least one subcategory. Notice that the result shows the correct item(s).
- Add space before or after the query, and notice that the result is still correct.

### Screenshots

https://user-images.githubusercontent.com/5533851/168208248-df7db941-672b-497a-b63a-816a32a2a708.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Categories/ProductCategoryListViewModelTests.swift [hunk 1]
```diff
@@ -183,6 +183,70 @@ final class ProductCategoryListViewModelTests: XCTestCase {
...
+        let viewModel = ProductCategoryListViewModel(siteID: siteID, storageManager: storageManager)
+        XCTAssertEqual(viewModel.categoryViewModels.count, 2) // confidence check
+
+        // When
+        viewModel.searchQuery = " western "
+
+        // Then
+        waitFor { promise in
+            // wait for 500 milliseconds due to debouncing
+            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
```
Comment (selanthiraiyan) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Categories/ProductCategoryListViewModelTests.swift]: Can we specify debounce time in a private property?

File: WooCommerce/Classes/ViewRelated/Products/Categories/ProductCategoryViewModelBuilder.swift [hunk 2]
```diff
@@ -43,6 +43,15 @@ extension ProductCategoryListViewModel {
             }
         }
 
+        /// Returns a simple array of `ProductCategoryCellViewModel` from the provided `categories` disregarding the parent-child relationships.
+        /// Provide an array of `selectedCategories` to properly reflect the selected state in the returned view model array.
+        ///
+        static func flatViewModels(from categories: [ProductCategory], selectedCategories: [ProductCategory]) -> [ProductCategoryCellViewModel] {
+            return categories.map { category in
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Products/Categories/ProductCategoryViewModelBuilder.swift]: What do you think about removing `return` here?

**Review by selanthiraiyan — APPROVED** [review]
Nice work @itsmeichigo 

It tests well. 🎉

File: WooCommerce/Classes/ViewRelated/Products/Categories/ProductCategoryViewModelBuilder.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Products/Categories/ProductCategoryViewModelBuilder.swift]: Done in 1f5beb3da5107bb09a4901788653d34eb4f306f4.

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Categories/ProductCategoryListViewModelTests.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/WooCommerceTests/ViewRelated/Products/Categories/ProductCategoryListViewModelTests.swift]: Done in c2330856c469e3d010818602efec18a5bbcf6cbb.

**Pull request merged**